Keep old database

jpgpi250 opened the pull request:
**By submitting this pull request, I confirm the following:** 
*please fill any appropriate checkboxes, e.g: [X]*

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
*A detailed description, screenshots (if necessary), as well as links to any relevant GitHub issues*
These changes will allow the user to keep, use and consult the old gravity database


**How does this PR accomplish the above?:**
*A detailed description (such as a changelog) and screenshots (if necessary) of the implemented fix*
option in setupVars.conf: KEEPOLDGRAVITY=true|false (default true)
pihole -g -s will switch the databases and restartdns. This will allow the user to check if things worked before the last gravity run
pihole -q -countnew will report the number of new entries, since the last gravity run
pihole -q -blocked will report the domains, new since the last gravity run, and already blocked a query

![image](https://user-images.githubusercontent.com/24669386/118379757-36663580-b5dd-11eb-8fa1-e4a958622ea9.png)

![image](https://user-images.githubusercontent.com/24669386/118379767-5138aa00-b5dd-11eb-8506-2793604fd788.png)

![image](https://user-images.githubusercontent.com/24669386/118379783-73322c80-b5dd-11eb-9dcf-a58ad68902b2.png)


**What documentation changes (if any) are needed to support this PR?:**
*A detailed list of any necessary changes*
setupVars.conf: KEEPOLDGRAVITY=true|false (default true)
new option for gravity (pihole -g -s OR pihole -g -switch)
new option for query (pihole -q -blocked)
new option for query 'pihole -q -countnew)

---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

jpgpi250 commented:
looks like it's not entirely correct again, DCO warning. I'm using github desktop, is there any way to achieve a correct signoff, using github desktop?
At least, I managed to solve the stickler warnings, using shellcheck...

dschaper commented:
Your signoff is good for me. I'll manually override that. You've included the line, DCObot wants a certain name and email pair but that's a bit picky.

jpgpi250 commented:
Thank you for fixing the DCO error, I was already preparing to close the PR and start again.

I've been at it for over 12 hours yesterday, thank you for supporting me in my first attempt to create a valid PR, thus hopefully contributing to pi-hole.

Unfortunately, I noticed an error in one of my echo commands, one to many ')' characters (at the end):
`  [i] Blocked: tdbtrk.com (file:///home/pi/quidsup/notrack-blocklist.txt))`

Now what do I do to correct this?
- start over again?
- a new commit of the file?
- an additional PR to correct this?

Sorry for the inconvenience, newbie, still learning...

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,68 @@ scanList(){
+      echo -e "  ${INFO} No new domains have been blocked yet."
+    else
+      for (( j=0; j<"${#resultarray[@]}"; j++ )); do
+        echo -e "  ${INFO} Blocked: ${resultarray[j]})"
```
```suggestion
        echo -e "  ${INFO} Blocked: ${resultarray[j]}"
```

This is the source of the extra `)`. You can even commit the change here on Github.

jpgpi250 replied:
Thank you, committed and signed (hopefully) the suggested change.

What would of happened if I edited the file, using github desktop, and committed? Would the PR automatically been updated?

yubiuser replied:
Yes. When you commit locally on your desktop and push the changes upstream to your github repo this PR will be updated automatically.

yubiuser commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,68 @@ scanList(){
+  newdomainentries=$(
+    sqlite3 "${gravityDBfile}" << EOSQL
+      ATTACH '${gravityOLDfile}' AS old;
+      SELECT count(DISTINCT domain) FROM gravity WHERE domain NOT IN (SELECT domain FROM old.gravity);
```
No tried, but it might be faster with an additional `distinct`

`SELECT count(DISTINCT domain) FROM gravity WHERE domain NOT IN (SELECT  DISTINCT domain FROM old.gravity);`

jpgpi250 replied:
without the additional distinct:
`time pihole -q -countnew
  [i] 44 new domain(s) found.

real    0m25.470s
user    0m16.324s
sys     0m5.009s
`

with the additional distinct:
` time pihole -q -countnew
  [i] 44 new domain(s) found.

real    1m13.115s
user    0m51.230s
sys     0m4.820s
`
I appreciate that you are reviewing this.

yubiuser replied:
Okay, let's forget that.

yubiuser commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,68 @@ scanList(){
+      JOIN adlist ON adlist.id = new.gravity.adlist_id
+      WHERE queries.domain NOT IN (SELECT domain FROM old.gravity)
+      AND queries.domain IN (SELECT domain FROM new.gravity)
+      AND queries.status IN (1,9)
```
In case of type 9, I think you need to take the domain from `additional_info` column because this is the actual blocked domain.

You might want to separate the CNAME stuff as it might get a bit to complexe otherwise.

jpgpi250 replied:
Would this be acceptable for you?
`pi@raspberrypi:~ $ pihole -q -blocked
  [i] Blocked: amzonse.xyz (https://phishing.army/download/phishing_army_blocklist_extended.txt)
  [i] Blocked: aod.cm.cn.miaozhen.com (https://www.github.developerdan.com/hosts/lists/ads-and-tracking-extended.txt)
  [i] Blocked: www.dnsimple.com. (CNAME)
        Domain: dnsimple.com (file:///home/pi/test.txt)
`
![image](https://user-images.githubusercontent.com/24669386/118401450-16c71f80-b666-11eb-8ea2-54a619480e13.png)

yubiuser replied:
Which domains is which? `www.dnsimple.com` is CNAME for `dnsimple.com`?

It wasn't meant as a hint for how to display things, but actually to include  domains from `additional_info` into your comparison. (You might have addressed this already and not just the output)

jpgpi250 replied:
I added a test list (could be a domain in a regular list) dnsimple.com. I than ran a dig www.dnsimple.com (not in any list). Since this query triggers CNAME detection (status 9), the separate query (as you suggested) finds www.dnsimple.com with status 9, so the user is informed the dns query for this domain, which is a CNAME was blocked. the second line (one for each CNAME) lists the gravity entry (dnsimple.com) and the list.

I think this should be very clear to the user, the output matches the info in the query log

![image](https://user-images.githubusercontent.com/24669386/118406829-e9866b80-b67d-11eb-8c63-4d0db4a0b91b.png)

yubiuser replied:
I've not checked the query in detail, but from a glance it looked like what I had in mind.

yubiuser commented on gravity.sh:
```diff
@@ -872,12 +889,44 @@ gravity_Cleanup() {
+      echo "         current database, updated $(date -d @"$(pihole-FTL "${gravityDBfile}" "SELECT value FROM info WHERE property IS 'updated';")")"
+      oldDBdate=$(pihole-FTL "${gravityOLDfile}" "SELECT value FROM info WHERE property IS 'updated';")
+      echo "         old database, updated $(date -d @"${oldDBdate}")"
+      mv "${gravityOLDfile}" "$(dirname -- "${gravityOLDfile}")/gravity_mv.db"
```
This might be a good place to add a check for disk space again?

jpgpi250 replied:
I'm open to all suggestions, but why should I check the diskspace here? No additional diskspace is required...
What happens is:
- rename gravity_old.db -> gravity_mv.db
- rename gravity.db -> gravity_old.db
- rename gravity_mv.db -> gravity.db
End result is a name switch, nothing more.
Are you suggesting to simply remove gravity_old.db, if there isn't enough diskspace, and thus, not perform a switch?

yubiuser replied:
You're right. I mixed it up.

DL6ER reviewed (commented):
How long do these queries run overall? It may be worth adding a message in the sense of "be patient, this may take some time" (better wording is welcome) if otherwise the user would sit like half minute asking themselves if it is still running or already dead.

jpgpi250 commented:
> How long do these queries run overall? It may be worth adding a message in the sense of "be patient, this may take some time" (better wording is welcome) if otherwise the user would sit like half minute asking themselves if it is still running or already dead.

Would this be acceptable to you?

![Animation](https://user-images.githubusercontent.com/24669386/118402716-6d832800-b66b-11eb-8d45-a7a9fe88decb.gif)

The query actually takes about 25 seconds (raspberry pi 3B).

dschaper commented:
If you dig around the code and history you might find a function we had called a spinner. @PromoFaux you think this may be a legitimate use of that spinner?

jpgpi250 commented:
Just pushed an update with the suggested changes, ready for review.

Thanks to all for the suggestions, feedback, time and effort.
Thanks for tolerating a newbie.

yubiuser commented:
> If you dig around the code and history you might find a function we had called a spinner.

(I was curious and found it)
https://github.com/pi-hole/pi-hole/pull/459

jpgpi250 commented:
at dschaper, PromoFaux, yubiuser.

Do I need to reintroduce the spinner function, or do you accept the solution I used.

May I suggest, If the spinner function is a requirement, to put this code in /opt/pihole/COL_TABLE (don't even know this will work, since it is a function - correct me if I'm wrong). It might be usable in other scripts...

DL6ER commented:
#459 said:

> It doesn't work when run by `cron`

It filled the log with `/\-/\-/\-/\-/\-/\-/\-/\-/\-/\-/\-/\-/\-/\- [...]` or something similar but this new feature should not be part of a `cron` job, anyway. I'm okay with the hint variant without any moving spinner element.

I yet have to find the time for a proper review, I only glanced through so far.

PromoFaux commented:
Yeah, no don't reintroduce that. It looked pretty, but it wasn't very reliable.

I'm good with a simple:

`  [i] Please wait...` 

or something like that

jpgpi250 commented:
- added 'please be patient' messages for all new queries I added
- added a new option listnew (pihole -q -listnew [-all]) the result is limited to 100 domains, unless '-all' is a command line option.

For my tests, in order to get results, with a lot of changes, I've added the list https://dbl.oisd.nl (yes I know, not your favorite list). This list changes every couple of hours, ensuring you get results (e.g. changes between the old and the new gravity database).

To test the CNAME code, I created a file, containing 'dnsimple.com', added it (file:///home/pi.test.txt) and ran 'pihole -g'. Than, dig 'www.dnsimple.com', this will cause a CNAME block, and the correct result, when running 'pihole -q - blocked'

jpgpi250 commented:
idea (NOT part of this PR)

The results of comparing the new and old gravity database appear to be correct.

I was testing the following:

add a field 'new_entry' to the gravity table of gravity.db, default 0
`ALTER TABLE gravity ADD COLUMN new_entry INTEGER DEFAULT 0;`

by running the following update on the database, while both old (temp) and new still exist, this field would be updated:
`( sudo sqlite3 "${gravityDBfile}" << EOSQL
      ATTACH '${gravityOLDfile}' AS old;
      UPDATE gravity SET new_entry = '1'
	  WHERE domain IN 
	    (SELECT domain FROM gravity WHERE domain NOT IN (SELECT domain FROM old.gravity));
EOSQL
      )` 
On a database with 4.087.232 entries and 4.223 new entries this takes only 35 seconds (raspberry pi 3B).
`real    0m34.983s
user    0m14.937s
sys     0m0.767s`

This database field could than be used to inform the user in the web interface, a new gravity entry caused a blocked query, for example add ' (NEW') behind the domain name.
This could be the user friendly way to inform the user blocking behavior has changed, due to a recent gravity update.

Unfortunately, I'm not smart / confident enough to update all the required components of pi-hole (API, web), thus asking for help.

DL6ER commented:
We should discuss the addition of a new database column somewhere else. Adding it to the `gravity` table does have some impact because it is the largest (or second largest behind `queries`) database table users will have. 35 seconds for 4 mio queries doesn't sound that dramatic. TBH, I'd have expected worse.

Points for the discussion somewhere else (likely Discourse):
- How long does gravity run overall with such a high number of domains? I'm wondering if we're talking about an increase of 1% or 10% or 50% in total runtime.
- How large is the same database once with and once without the extra column? It'll be surely some megabytes of difference.

This would also need a change in FTL. Currently, we're are just using the tree index of gravity when checking if a domain is blocked. This index is loaded into memory and super-fast even when there are dozens of million of domains in gravity. However, when this change is made, FTL needs to actually look up the database record to get the value of `new`. Surely doable but it'd remove one major speed benefit of the database. I'm neither convinced that it worth it nor do I have any actual numbers for comparison at hand.

jpgpi250 commented:
Where do you want to discuss it than? [Here](https://discourse.pi-hole.net/t/gravity-database/46182)?

I can already answer some of your question regarding this, I've been experimenting already with this.

remember, my gravity database has 4.087.232 entries, there were 4.223 unique new entries, compared to the previous gravity database (I keep an old version of the gravity database, this to increase the diff between the databases every time I run 'pihole -g' (manual copy required)

![image](https://user-images.githubusercontent.com/24669386/119223871-17dfcd00-bafc-11eb-905b-e18190233f46.png)

How long does gravity run? As I see it, this wouldn't affect the gravity run at all (it would require new entries to be added with 3 fields, currently 2, but the new field would always be '0').
The gravity build would't change at all. The idea is to modify the 'new_entry' column (run the update query) just before the gravity.sh instruction 'mv "${gravityTEMPfile}" "${gravityDBfile}"' (apply the change on the 'TEMPfile'). This would than add a minute, give or take to the gravity run, but since that happens at night (sunday morning), you would only notice this if you forced a run, using 'pihole -g'

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,131 @@ scanList(){
+CountNewDomains() {
+  printf %s "  ${INFO} Running query, please be patient..."
+  # store query result (count) in variable
+  newdomainentries=$(
```
```suggestion
  newdomainentries="$(
```

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,131 @@ scanList(){
+      ATTACH '${gravityOLDfile}' AS old;
+      SELECT count(DISTINCT domain) FROM gravity WHERE domain NOT IN (SELECT domain FROM old.gravity);
+EOSQL
+      )
```
```suggestion
      )"
```

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,131 @@ scanList(){
+  mapfile -t newdomainarray < <( sqlite3 "${gravityDBfile}" << EOSQL
+      ATTACH '${gravityOLDfile}' AS old;
+      SELECT domain, '' || adlist.address || '' FROM gravity
+      JOIN adlist ON adlist.id = gravity.adlist_id	  
```
```suggestion
      JOIN adlist ON adlist.id = gravity.adlist_id
```

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,131 @@ scanList(){
+
+# function to list new domain entries since last gravity run
+ListNewDomains() {
+  if [ -z "${all}" ]; then
+    QueryLimit="Limit 100"
+  fi
```
```suggestion
  QueryLimit=""
  if [ -z "${all}" ]; then
    QueryLimit="Limit 100"
  fi
```

To prevent a possible SQL injection from outside. This is easily demonstrated by:
``` bash
echo "QueryLimit=\"ABC\"" >> /etc/pihole/pihole-FTL.conf
pihole -q -listnew -all
```
and getting the error:
``` plain
  [i] Running query, please be patient...Error: near line 2: near "ABC": syntax error
  [i] No new domains have been found.
```

The injection could have been something worse, like disabling all adlists or injecting some (unwanted) whitelist domains.

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,131 @@ scanList(){
+    echo -e "  ${domain} (${COL_GREEN}${urllist}${COL_NC})"
+  done
+  if [[ -z "${all}" ]] && [[ "${#newdomainarray[*]}" -ge 100 ]]; then
+    echo -e "  ${INFO} Over 100 new domains found.
+        This can be overridden using the -all option"
```
Clarify what can be overwritten: the limit.

```suggestion
    echo -e "  ${INFO} Over 100 new domains found.
        If you want to see them all, append \"-all\" to the command"
```

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -58,12 +61,131 @@ scanList(){
 Query the adlists for a specified domain
 
 Options:
+  -blocked            List blocked domains since last gravity run
```
I wonder if the command should also have a trailing `new` as the other have. Just for consistency. This is open for debate, of course.
```suggestion
  -blockednew         List blocked domains since last gravity run
```

DL6ER commented on gravity.sh:
```diff
@@ -73,6 +73,7 @@ fi
 # have changed
 gravityDBfile="${GRAVITYDB}"
 gravityTEMPfile="${GRAVITYDB}_temp"
+gravityOLDfile="$(dirname -- "${gravityDBfile}")/gravity_old.db"
```
```suggestion
gravityDIR="$(dirname -- "${gravityDBfile}")"
gravityOLDfile="${gravityDIR}/gravity_old.db"
```

DL6ER commented on gravity.sh:
```diff
@@ -123,8 +124,24 @@ gravity_swap_databases() {
+
+  # Swap databases and remove or conditionally rename old database
+  # Number of available blocks on disk
+  availableBlocks=$(stat -f --format "%a" "$(dirname -- "${gravityDBfile}")")
```
```suggestion
  availableBlocks=$(stat -f --format "%a" "${gravityDIR}")
```

DL6ER commented on gravity.sh:
```diff
@@ -123,8 +124,24 @@ gravity_swap_databases() {
+  gravityBlocks=$(stat --format "%b" ${gravityDBfile})
+  # Only keep the old database if available disk space is at least twice the size of the existing gravity.db.
+  # Better be safe than sorry...
+  # The variable KEEPOLDGRAVITY in /etc/pihole-FTL.conf can be used to disable this feature.
```
```suggestion
  # The variable KEEPOLDGRAVITY in /etc/pihole/pihole-FTL.conf can be used to disable this feature.
```

DL6ER commented on gravity.sh:
```diff
@@ -872,12 +889,44 @@ gravity_Cleanup() {
+      echo "         current database, updated $(date -d @"$(pihole-FTL "${gravityDBfile}" "SELECT value FROM info WHERE property IS 'updated';")")"
+      oldDBdate=$(pihole-FTL "${gravityOLDfile}" "SELECT value FROM info WHERE property IS 'updated';")
+      echo "         old database, updated $(date -d @"${oldDBdate}")"
+      mv "${gravityOLDfile}" "$(dirname -- "${gravityOLDfile}")/gravity_mv.db"
```
```suggestion
      mv "${gravityOLDfile}" "${gravityDIR}/gravity_mv.db"
```

DL6ER commented on gravity.sh:
```diff
@@ -872,12 +889,44 @@ gravity_Cleanup() {
+      echo "         old database, updated $(date -d @"${oldDBdate}")"
+      mv "${gravityOLDfile}" "$(dirname -- "${gravityOLDfile}")/gravity_mv.db"
+      mv "${gravityDBfile}" "${gravityOLDfile}"
+      mv "$(dirname -- "${gravityOLDfile}")/gravity_mv.db" "${gravityDBfile}"
```
```suggestion
      mv "${gravityDIR}/gravity_mv.db" "${gravityDBfile}"
```

DL6ER commented on advanced/Scripts/query.sh:
```diff
@@ -27,6 +28,8 @@ fi
 # Set this only after sourcing pihole-FTL.conf as the gravity database path may
 # have changed
 gravityDBfile="${GRAVITYDB}"
+gravityOLDfile="$(dirname -- "${gravityDBfile}")/gravity_old.db"
```
```suggestion
gravityDIR="$(dirname -- "${gravityDBfile}")"
gravityOLDfile="${gravityDIR}/gravity_old.db"
```

DL6ER reviewed (changes requested):
Thanks for your submission and sorry for the review having taken so long. This is almost good to go, there are a few things I've found, including one SQL code injection vulnerability that needs to be fixed.

yubiuser commented:
> setupVars.conf: KEEPOLDGRAVITY=true|false (default true)

I don't think this should be `true` by default. If users are so advanced that they know/want/need this possibility they can easily switch it on.

jpgpi250 commented:
I applied all suggested changes and some...
- query.sh : **blocked** needed to be changed into **blockednew** twice
- gravity.sh: eliminating excessive use of **dirname --** required adding **gravityDIR="$(dirname -- "${gravityDBfile}")"**. This was only suggested in query.sh

I don't understand why **rm "${gravityDBfile}"** should be removed. It's part of the original code. What I changed is:
- rename, using mv, DB to DB_old (if the conditions are ok, e.g. diskcheck and KEEPOLDGRAVITY=true)
OR
- erase, using rm, the DB (this is required because the next step is rename, using mv, DBtemp to DB.
The only other place where the identical rm command is used, is in the recreate_database code.

jpgpi250 commented:
> `I don't think this should be true by default. If users are so advanced that they know/want/need this possibility they can easily switch it on.`

This will require some extra code to hide the additional options (help), and block processing of the command line options, if the value is false.

Fine by me, if everybody thinks this is necessary, just let me know...

jfb-pihole commented:
I agree that the default should be false.

jpgpi250 commented:
> I agree that the default should be false.

> I don't think this should be true by default. If users are so advanced that they know/want/need this possibility they can easily switch it on.

done, default is false, additional help is hidden, new switches will not work, however, if KEEPOLDDATABASE is unset or false, the switches (-countnew, -listnew, -blockednew) will be interpreted as domain search parameter for pihole -q (searches for domain matches).

dschaper commented:
Was there a reason you closed and deleted the branch?

jpgpi250 commented:
Last Saturday, I used 'pihole -g -s' (switch databases) to check something, and forgot to switch them back. As a result, the cron gravityupdate, running on sunday morning, deleted the most recent database (not active at that time, but renamed to old), removing all changes to clients, whitelist, blacklist, regex, url's, I made last week.
My bad, unlucky, but since this might happen to other users, once this would be merged, it didn't feel like a good idea anymore, so I decided to remove everything.

I will be starting over, submit a new PR, keeping only the code to keep the old database, nothing more (no switching, no additional queries). Since the general opinion appears to be this is for advanced users only (default false), they can write / execute queries for themselves, if required.

dschaper commented:
Oh, completely understandable. Thanks for letting us know what is going on. We were concerned that maybe you closed it because we hadn't been active with it. We've all been swamped with non Pi-hole things so we weren't intentionally ignoring it.

Please do submit a new PR when you have something and we can talk it over.

jpgpi250 commented:
Sorry for the troubles I'm causing, thanks to all for tolerating me and my poor attempts to contribute.

dschaper commented:
Honestly those are not things that even came to mind. No troubles and your contributions are not poor.
